src/schedule/zephyr_domain.c

```diff
@@ -221,9 +246,38 @@ static int zephyr_domain_unregister(struct ll_schedule_domain *domain,
 	return 0;
 }
 
+#if CONFIG_CROSS_CORE_STREAM
+static void zephyr_domain_block(struct ll_schedule_domain *domain)
+{
+	struct zephyr_domain *zephyr_domain = ll_sch_domain_get_pdata(domain);
+
+	tr_dbg(&ll_tr, "Blocking LL scheduler");
+
+	k_mutex_lock(&zephyr_domain->block_mutex, K_FOREVER);
+	atomic_set(&zephyr_domain->block, 1);
```

Comment:
Btw, does this have to be atomic since all users should need to gain the mutex before use ? i.e. we could save some cycles with a regular int ?

Reply:
It has to be volatile to make sure compiler does not optimize reading from memory when checking block value in second "if".

```
if (atomic_get(&zephyr_domain->block)) {
	k_mutex_lock(&zephyr_domain->block_mutex, K_FOREVER);
	if (atomic_get(&zephyr_domain->block))
		k_condvar_wait(&zephyr_domain->block_condvar,
			       &zephyr_domain->block_mutex, K_FOREVER);
	k_mutex_unlock(&zephyr_domain->block_mutex);
}
```
Without volatile compiler may optimize second read of zephyr_domain->block but we want to actually check block value after locking the mutex.

atomic here is used not for the purpose of atomic read/write, but for a purpose of volatile read/write. I could replace atomic with volatile bool, however, we seems use atomic in similar cases in SOF.

Reply:
ok, but should't the code block be (if I'm understanding the locking around ```zephyr_domain->block```)
```c
/* get the global domain mutex */
k_mutex_lock(&zephyr_domain->block_mutex, K_FOREVER);

/* now critical section with lock held - no one should read/write domain->block without holding mutex */
if (zephyr_domain->block) /* can now be an int since lock is held and no one should change without mtex held */
	k_condvar_wait(&zephyr_domain->block_condvar,
		       &zephyr_domain->block_mutex, K_FOREVER);

/* leave the critical section */
k_mutex_unlock(&zephyr_domain->block_mutex);
```